contracts/cw-governance/src/contract.rs

```diff
@@ -0,0 +1,359 @@
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const GOV_MODULE_REPLY_ID: u64 = 0;
+const VOTE_MODULE_INSTANTIATE_REPLY_ID: u64 = 1;
+const VOTE_MODULE_UPDATE_REPLY_ID: u64 = 2;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let config = Config {
+        name: msg.name,
+        description: msg.description,
+        image_url: msg.image_url,
+    };
+    CONFIG.save(deps.storage, &config)?;
+
+    let vote_module_msg = msg
+        .voting_module_instantiate_info
+        .into_wasm_msg(env.contract.address.clone());
+    let vote_module_msg: SubMsg<Empty> =
+        SubMsg::reply_on_success(vote_module_msg, VOTE_MODULE_INSTANTIATE_REPLY_ID);
+
+    let gov_module_msgs: Vec<SubMsg<Empty>> = msg
+        .governance_modules_instantiate_info
+        .into_iter()
+        .map(|info| info.into_wasm_msg(env.contract.address.clone()))
+        .map(|wasm| SubMsg::reply_on_success(wasm, GOV_MODULE_REPLY_ID))
+        .collect();
+    if gov_module_msgs.is_empty() {
+        return Err(ContractError::NoGovernanceModule {});
+    }
+
+    GOVERNANCE_MODULE_COUNT.save(deps.storage, &(gov_module_msgs.len() as u64))?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("sender", info.sender)
+        .add_submessage(vote_module_msg)
+        .add_submessages(gov_module_msgs))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::ExecuteProposalHook { .. } => {
+            if !GOVERNANCE_MODULES.has(deps.storage, info.sender.clone()) {
+                return Err(ContractError::Unauthorized {});
+            }
+        }
+        _ => {
+            if info.sender != env.contract.address {
+                return Err(ContractError::Unauthorized {});
+            }
+        }
+    }
+
+    let response = match msg {
+        ExecuteMsg::ExecuteProposalHook { msgs } => execute_proposal_hook(msgs),
+        ExecuteMsg::UpdateConfig { config } => execute_update_config(deps, config),
+        ExecuteMsg::UpdateVotingModule { module } => execute_update_voting_module(env, module),
+        ExecuteMsg::UpdateGovernanceModules { to_add, to_remove } => {
+            execute_update_governance_modules(deps, env, to_add, to_remove)
+        }
+        ExecuteMsg::SetItem { key, addr } => execute_set_item(deps, key, addr),
+        ExecuteMsg::RemoveItem { key } => execute_remove_item(deps, key),
+    }?;
+
+    Ok(response.add_attribute("sender", info.sender))
+}
+
+pub fn execute_proposal_hook(msgs: Vec<CosmosMsg<Empty>>) -> Result<Response, ContractError> {
+    Ok(Response::default()
```

Comment:
I could still not wrap the design around my head, but reflecting the proposal hook without any auth looks scary at first sight.

Reply:
There is validation above that checks that this is coming from a governance module in the execute method. I think it makes sense to move that into the function though for future readers though.